src/Confluent.Kafka/IProducer.cs

```diff
@@ -0,0 +1,149 @@
+        /// <include file='include_docs_producer.xml' path='API/Member[@name="ProduceAsync_Common"]/*' />
+        Task<Message> ProduceAsync(string topic, byte[] key, byte[] val);
+
+        /// <include file='include_docs_producer.xml' path='API/Member[@name="ProduceAsync_string_Partition_TKey_TValue_Timestamp_IEnumerable"]/*' />
+        /// <include file='include_docs_producer.xml' path='API/Member[@name="ProduceAsync_Common"]/*' />
+        Task<Message> ProduceAsync(
+            string topic, Partition partition, 
+            byte[] key, byte[] val, 
+            Timestamp timestamp, 
+            IEnumerable<KeyValuePair<string, byte[]>> headers);
+
+        /// <include file='include_docs_producer.xml' path='API/Member[@name="ProduceAsync_string_Partition_byte_int_int_byte_int_int_Timestamp_IEnumerable"]/*' />
+        /// <include file='include_docs_producer.xml' path='API/Member[@name="ProduceAsync_Common"]/*' />
+        Task<Message> ProduceAsync(
+            string topic, Partition partition, 
+            byte[] key, int keyOffset, int keyLength, 
+            byte[] val, int valOffset, int valLength, 
+            Timestamp timestamp, IEnumerable<KeyValuePair<string, byte[]>> headers);
+
+        /// <include file='include_docs_producer.xml' path='API/Member[@name="ProduceAsync_Message"]/*' />
+        /// <include file='include_docs_producer.xml' path='API/Member[@name="Produce_IDeliveryHandler"]/*' />
+        void Produce(Message message, IDeliveryHandler deliveryHandler);
+
+        /// <include file='include_docs_producer.xml' path='API/Member[@name="ProduceAsync_string_TKey_TValue"]/*' />
+        /// <include file='include_docs_producer.xml' path='API/Member[@name="Produce_IDeliveryHandler"]/*' />
+        void Produce(string topic, byte[] key, byte[] val, IDeliveryHandler deliveryHandler);
+
+        /// <include file='include_docs_producer.xml' path='API/Member[@name="ProduceAsync_string_Partition_TKey_TValue_Timestamp_IEnumerable"]/*' />
+        /// <include file='include_docs_producer.xml' path='API/Member[@name="Produce_IDeliveryHandler"]/*' />
+        void Produce(
+            string topic, Partition partition,
+            byte[] key, byte[] val,
+            Timestamp timestamp,
+            IEnumerable<KeyValuePair<string, byte[]>> headers,
+            IDeliveryHandler deliveryHandler);
+
+        /// <include file='include_docs_producer.xml' path='API/Member[@name="ProduceAsync_string_Partition_byte_int_int_byte_int_int_Timestamp_IEnumerable"]/*' />        
+        /// <include file='include_docs_producer.xml' path='API/Member[@name="Produce_IDeliveryHandler"]/*' />
+        void Produce(
+            string topic, Partition partition, 
+            byte[] key, int keyOffset, int keyLength,
+            byte[] val, int valOffset, int valLength,
+            Timestamp timestamp, IEnumerable<KeyValuePair<string, byte[]>> headers,
```

Comment:
C# people are used to reading `KeyValuePair<blah>`, so it doesn't look as ugly to us as it does to you (it looks nice and standard actually), but I suspect your point on hiding the header structure is a very good one.

Note that `IEnumerable<Header>` is not compatible with `IEnumerable<KeyValuePair<string, byte[]>` (even if we had a `Header` type that derived from `KeyValuePair<string, byte[]>` which we don't and also can't since `KeyValuePair` is sealed).

So I think we ditch `KeyValuePair<string,byte[]>` altogether and replace it with a new `Header` class that doesn't derive from anything. This gives the same scope for future enhancement as the Java API.

Reply:
Another advantage of `KeyValuePair<string,byte[]>` that I was forgetting about and was somewhat driving my decision here is `IEnumerable<KeyValuePair<string,byte[]>` is compatible with `Dictionary<string, byte[]>`. This is not hugely important though - your argument for introducing a `Header` class is still winning.